src/java/org/apache/cassandra/index/sai/disk/v1/bbtree/BlockBalancedTreeWriter.java

```diff
@@ -0,0 +1,779 @@
+
+            treeOutput.writeVInt(leafValueCount);
+
+            for (int x = 0; x < leafValueCount; x++)
+            {
+                rowIDAndIndexes[x].valueOrderIndex = x;
+                rowIDAndIndexes[x].rowID = leafRowIDs[x];
+            }
+
+            final Sorter sorter = new IntroSorter()
+            {
+                RowIDAndIndex pivot;
+
+                @Override
+                protected void swap(int i, int j)
+                {
+                    RowIDAndIndex o = rowIDAndIndexes[i];
+                    rowIDAndIndexes[i] = rowIDAndIndexes[j];
+                    rowIDAndIndexes[j] = o;
+                }
+
+                @Override
+                protected void setPivot(int i)
+                {
+                    pivot = rowIDAndIndexes[i];
+                }
+
+                @Override
+                protected int comparePivot(int j)
+                {
+                    return Long.compare(pivot.rowID, rowIDAndIndexes[j].rowID);
+                }
+            };
+
+            sorter.sort(0, leafValueCount);
+
+            // write leaf rowID -> orig index
+            scratchOut2.reset();
+
+            // iterate in row ID order to get the row ID index for the given value order index
+            // place into an array to be written as packed ints
+            for (int x = 0; x < leafValueCount; x++)
+                orderIndex[rowIDAndIndexes[x].valueOrderIndex] = x;
+
+            LeafOrderMap.write(orderIndex, leafValueCount, maxPointsInLeafNode - 1, scratchOut2);
+
+            treeOutput.writeVInt(scratchOut2.getPosition());
+            treeOutput.writeBytes(scratchOut2.getBytes(), 0, scratchOut2.getPosition());
+
+            callback.writeLeafDocs(rowIDAndIndexes, 0, leafValueCount);
+
+            writeCommonPrefix(scratchOut, commonPrefixLength, leafValues);
+
+            scratchBytesRef1.length = bytesPerValue;
+            scratchBytesRef1.bytes = leafValues;
+
+            IntFunction<BytesRef> packedValues = (i) -> {
+                scratchBytesRef1.offset = bytesPerValue * i;
+                return scratchBytesRef1;
+            };
+            if (DEBUG)
+                valuesInOrderAndBounds(leafValueCount,
+                                       ArrayUtil.copyOfSubArray(leafValues, 0, bytesPerValue),
+                                       ArrayUtil.copyOfSubArray(leafValues, (leafValueCount - 1) * bytesPerValue, leafValueCount * bytesPerValue),
+                                       packedValues,
+                                       leafRowIDs);
+
+            writeLeafBlockPackedValues(scratchOut, commonPrefixLength, leafValueCount, packedValues);
+
+            treeOutput.writeBytes(scratchOut.getBytes(), 0, scratchOut.getPosition());
+
+            scratchOut.reset();
+        }
+
+        private void checkMaxLeafNodeCount(int numLeaves)
+        {
+            if ((1 + bytesPerValue) * (long) numLeaves > ArrayUtil.MAX_ARRAY_LENGTH)
```

Comment:
Where does the `1 + ` come from?

Reply:
It's wrong and I removed it. Originally the packed index contained a dimension for every split value but we don't need that because we have a single dimension.